File: renderer/src/test/java/io/sparkled/util/LedTestUtils.java
package io.sparkled.util;

import io.sparkled.model.render.Led;
import io.sparkled.model.render.RenderedChannel;
import io.sparkled.model.render.RenderedFrame;

import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * Convenience methods for working with LEDs in tests.
 */
public class LedTestUtils {
    private LedTestUtils() {
    }

    public static String toLedString(RenderedChannel channel) {
        return channel.getFrames()
                .stream()
                .map(LedTestUtils::toLedString)
                .collect(Collectors.joining("\n"));
    }

    public static String toLedString(int[][] leds) {
        return Arrays.stream(leds)
                .map(LedTestUtils::toLedString)
                .collect(Collectors.joining("\n"));
    }

    public static String toLedString(RenderedFrame frame) {
        return IntStream.range(0, frame.getLedCount())
                .mapToObj(frame::getLed)
                .map(Led::toString)
                .collect(Collectors.joining(", "));
    }

    public static String toLedString(int[] leds) {
        return Arrays.stream(leds)
                .mapToObj(LedTestUtils::getLedFromRgb)
                .map(Led::toString)
                .collect(Collectors.joining(", "));
    }

    private static Led getLedFromRgb(int rgb) {
        byte[] ledData = new byte[3];
        Led led = new Led(ledData, 0, 0);
        led.addRgb((rgb & 0xFF0000) >> 16, (rgb & 0x00FF00) >> 8, (rgb & 0x0000FF));
        return led;
    }
}

File: renderer/src/test/kotlin/io/sparkled/util/LedTestUtils.kt
package io.sparkled.util

import io.sparkled.model.render.Led
import io.sparkled.model.render.RenderedFrame
import io.sparkled.model.render.RenderedStagePropData

/**
 * Convenience methods for working with LEDs in tests.
 */
object LedTestUtils {

    fun toLedString(channel: RenderedStagePropData): String {
        return channel.frames.map { toLedString(it) }.joinToString("\n")
    }

    fun toLedString(leds: Array<IntArray>): String {
        return leds.map { toLedString(it) }.joinToString("\n")
    }

    fun toLedString(frame: RenderedFrame): String {
        return (0 until frame.ledCount).map { frame.getLed(it) }.joinToString(", ", transform = Led::toString)
    }

    fun toLedString(leds: IntArray): String {
        return leds.map(this::getLedFromRgb).joinToString(", ", transform = Led::toString)
    }

    private fun getLedFromRgb(rgb: Int): Led {
        val ledData = ByteArray(3)
        val led = Led(ledData, 0, 0)
        led.addRgb(rgb and 0xFF0000 shr 16, rgb and 0x00FF00 shr 8, rgb and 0x0000FF)
        return led
    }
}
